### ISSUE TITLE
Running into an issue with evaluating the model

### ISSUE THREAD

**Issue by jaddoescad** [aws-samples/amazon-sagemaker-pytorch-detectron2#1]
I'm running your example as a demo and it keeps failing whenever I introduce evaluation to the training.

I get the following error:
```
TypeError: _eval_predictions() missing 1 required positional argument: 'predictions'

Traceback (most recent call last):File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrapfn(i, *args)File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/launch.py", line 94, in _distributed_workermain_func(*args)File "/opt/ml/code/training.py", line 172, in _train_implTrainer.test(cfg, model, evaluator)File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/defaults.py", line 552, in testresults_i = inference_on_dataset(model, data_loader, evaluator)File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/evaluator.py", line 182, in inference_on_datasetresults = evaluator.evaluate()File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/coco_evaluation.py", line 175, in evaluateself._eval_predictions(predictions, img_ids=img_ids)
```
--

Have you ran into this issue?

**Comment by levon-khojoyan** [comment]
Hi @jaddoescad ! were you able to solve the issue?

**Comment by andrey-batutin-da** [comment]
having same issue for me

**Comment by paoloirrera** [comment]
Hello, I have just executed the training session of the notebook demo and the training job was successful. Could please share here your changes if any and the CloudWatch log of the failing job?

**Comment by levon-khojoyan** [comment]
2021-04-26T22:31:48.654+03:00 | #011nonzero(*, bool as_tuple) (Triggered internally at /pytorch/torch/csrc/utils/python_arg_parser.cpp:766.) num_fg = fg_inds.nonzero().numel()
-- | --
  | 2021-04-26T22:31:48.654+03:00 | /opt/conda/lib/python3.6/site-packages/detectron2/modeling/roi_heads/fast_rcnn.py:103: UserWarning: This overload of nonzero is deprecated:
  | 2021-04-26T22:31:48.654+03:00 | #011nonzero()
  | 2021-04-26T22:31:48.654+03:00 | Consider using one of the following signatures instead:
  | 2021-04-26T22:31:48.654+03:00 | #011nonzero(*, bool as_tuple) (Triggered internally at /pytorch/torch/csrc/utils/python_arg_parser.cpp:766.) return start_processes(fn, args, nprocs, join, daemon, start_method='spawn') File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 158, in start_processes while not context.join(): File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 119, in join raise Exception(msg)
  | 2021-04-26T22:31:48.654+03:00 | Exception:
  | 2021-04-26T22:31:48.654+03:00Copy-- Process 0 terminated with the following error: | -- Process 0 terminated with the following error:
  | 2021-04-26T22:31:48.654+03:00CopyTraceback (most recent call last):  File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrap    fn(i, *args)  File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/launch.py", line 94, in _distributed_worker    main_func(*args)  File "/opt/ml/code/training.py", line 172, in _train_impl    Trainer.test(cfg, model, evaluator)  File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/defaults.py", line 552, in test    results_i = inference_on_dataset(model, data_loader, evaluator)  File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/evaluator.py", line 182, in inference_on_dataset    results = evaluator.evaluate()  File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/coco_evaluation.py", line 175, in evaluate    self._eval_predictions(predictions, img_ids=img_ids) | Traceback (most recent call last): File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrap fn(i, *args) File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/launch.py", line 94, in _distributed_worker main_func(*args) File "/opt/ml/code/training.py", line 172, in _train_impl Trainer.test(cfg, model, evaluator) File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/defaults.py", line 552, in test results_i = inference_on_dataset(model, data_loader, evaluator) File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/evaluator.py", line 182, in inference_on_dataset results = evaluator.evaluate() File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/coco_evaluation.py", line 175, in evaluate self._eval_predictions(predictions, img_ids=img_ids)
  | 2021-04-26T22:31:48.654+03:00CopyTypeError: _eval_predictions() missing 1 required positional argument: 'predictions'  num_fg = fg_inds.nonzero().numel() | TypeError: _eval_predictions() missing 1 required positional argument: 'predictions' num_fg = fg_inds.nonzero().numel()
  | 2021-04-26T22:31:48.654+03:00Copy/opt/conda/lib/python3.6/site-packages/detectron2/modeling/roi_heads/fast_rcnn.py:103: UserWarning: This overload of nonzero is deprecated: | /opt/conda/lib/python3.6/site-packages/detectron2/modeling/roi_heads/fast_rcnn.py:103: UserWarning: This overload of nonzero is deprecated:
  | 2021-04-26T22:31:48.654+03:00Copy#011nonzero() | #011nonzero()
  | 2021-04-26T22:31:48.654+03:00CopyConsider using one of the following signatures instead: | Consider using one of the following signatures instead:
  | 2021-04-26T22:31:48.654+03:00Copy#011nonzero(*, bool as_tuple) (Triggered internally at  /pytorch/torch/csrc/utils/python_arg_parser.cpp:766.)  num_fg = fg_inds.nonzero().numel() | #011nonzero(*, bool as_tuple) (Triggered internally at /pytorch/torch/csrc/utils/python_arg_parser.cpp:766.) num_fg = fg_inds.nonzero().numel()
  | 2021-04-26T22:31:48.655+03:00CopyTraceback (most recent call last):  File "training.py", line 459, in <module>    train(_parse_args())  File "training.py", line 206, in train    args=(args,),  File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/launch.py", line 59, in launch    daemon=False,  File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 200, in spawn    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')  File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 158, in start_processes    while not context.join():  File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 119, in join    raise Exception(msg) | Traceback (most recent call last): File "training.py", line 459, in <module> train(_parse_args()) File "training.py", line 206, in train args=(args,), File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/launch.py", line 59, in launch daemon=False, File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 200, in spawn return start_processes(fn, args, nprocs, join, daemon, start_method='spawn') File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 158, in start_processes while not context.join(): File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 119, in join raise Exception(msg)
  | 2021-04-26T22:31:48.655+03:00CopyException: | Exception:
  | 2021-04-26T22:31:48.655+03:00 | -- Process 0 terminated with the following error:
  | 2021-04-26T22:31:48.655+03:00CopyTraceback (most recent call last):  File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrap    fn(i, *args)  File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/launch.py", line 94, in _distributed_worker    main_func(*args)  File "/opt/ml/code/training.py", line 172, in _train_impl    Trainer.test(cfg, model, evaluator)  File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/defaults.py", line 552, in test    results_i = inference_on_dataset(model, data_loader, evaluator)  File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/evaluator.py", line 182, in inference_on_dataset    results = evaluator.evaluate()  File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/coco_evaluation.py", line 175, in evaluate    self._eval_predictions(predictions, img_ids=img_ids) | Traceback (most recent call last): File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrap fn(i, *args) File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/launch.py", line 94, in _distributed_worker main_func(*args) File "/opt/ml/code/training.py", line 172, in _train_impl Trainer.test(cfg, model, evaluator) File "/opt/conda/lib/python3.6/site-packages/detectron2/engine/defaults.py", line 552, in test results_i = inference_on_dataset(model, data_loader, evaluator) File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/evaluator.py", line 182, in inference_on_dataset results = evaluator.evaluate() File "/opt/conda/lib/python3.6/site-packages/detectron2/evaluation/coco_evaluation.py", line 175, in evaluate self._eval_predictions(predictions, img_ids=img_ids)
  | 2021-04-26T22:31:48.655+03:00CopyTypeError: _eval_predictions() missing 1 required positional argument: 'predictions' | TypeError: _eval_predictions() missing 1 required positional argument: 'predictions'

**Comment by levon-khojoyan** [comment]
The only change is the augmented training set, the rest are default, I have tried both the 'fast' and 'coco' evaluations, it calculates the val losses and fails the job.
We've also built and pushed the image from EC2, as it was throwing insufficient memory issues on sagemaker.

**Comment by paoloirrera** [comment]
The release of detectron2 version 0.4 upgrade caused evaluation to break because the signature of _eval_predictions changed. I solved in commit [e65a32e](https://github.com/aws-samples/amazon-sagemaker-pytorch-detectron2/commit/e65a32e2490a75fe9276e5c86a556938be83678e):

- Update the signature of D2CocoEvaluator._eval_predictions
- Specify Detectron2 0.4 in Docker files

As for memory issues in SageMaker, you can increase the size of the EBS volume attached to your notebook instance.
